test/test_python_factory.py

```diff
@@ -0,0 +1,173 @@
+                    "targets": {
+                        "1": {
+                            "module": "wsgi",
+                            "callable": "wsgi_a",
+                            "factory": False
+                        },
+                        "2": {
+                            "module": "wsgi",
+                            "callable": "wsgi_b",
+                            "factory": False
+                        },
+                        "factory-1": {
+                            "module": "wsgi",
+                            "callable": "wsgi_a_factory",
+                            "factory": True
+                        },
+                        "factory-2": {
+                            "module": "wsgi",
+                            "callable": "wsgi_b_factory",
+                            "factory": True
+                        },
+                    },
+                }
+            },
+        }
+    )
+
+    resp = client.get(url='/1')
+    assert resp['status'] == 200
+    assert resp['body'] == '1'
+
+    resp = client.get(url='/2')
+    assert resp['status'] == 200
+    assert resp['body'] == '2'
+
+    resp = client.get(url='/factory-1')
+    assert resp['status'] == 200
+    assert resp['body'] == '1'
+
+    resp = client.get(url='/factory-2')
+    assert resp['status'] == 200
+    assert resp['body'] == '2'
+
+def test_python_factory_without_targets():
+    python_dir = f'{option.test_dir}/python'
+
+    assert 'success' in client.conf(
+        {
+            "listeners": {"*:8080": {"pass": "routes"}},
+            "routes": [
+                {
+                    "match": {"uri": "/1"},
+                    "action": {"pass": "applications/python-app-factory"},
+                },
+                {
+                    "match": {"uri": "/2"},
+                    "action": {"pass": "applications/python-app"},
+                }
+            ],
+            "applications": {
+                "python-app-factory": {
+                    "type": client.get_application_type(),
+                    "working_directory": f'{python_dir}/factory/',
+                    "path": f'{python_dir}/factory/',
+                    "module": "wsgi",
+                    "callable": "wsgi_a_factory",
+                    "factory": True
+                },
+                "python-app": {
+                    "type": client.get_application_type(),
+                    "working_directory": f'{python_dir}/factory/',
+                    "path": f'{python_dir}/factory/',
+                    "module": "wsgi",
+                    "callable": "wsgi_b",
+                    "factory": False
+                }
+            },
+        }
+    )
+
+    resp = client.get(url='/1')
+    assert resp['status'] == 200
+    assert resp['body'] == '1'
+
+    resp = client.get(url='/2')
+    assert resp['status'] == 200
+    assert resp['body'] == '2'
+
+def test_python_factory_returns_invalid_callable():
```

Comment:
`test_python_factory_returns_invalid_callable` and `test_python_factory_is_invalid_callable` are only 1 line diff. So I would recommend to combine them and avoid repeating if possible.